src/shaders.rs

```diff
@@ -305,38 +298,47 @@ pub fn full_shaders(device: &Device, engine: &mut WgpuEngine) -> Result<FullShad
             BindType::Buffer,
         ],
     )?;
-    let fine = match ANTIALIASING {
-        crate::AaConfig::Area => engine.add_shader(
-            device,
-            "fine",
-            preprocess::preprocess(shader!("fine"), &full_config, &imports).into(),
-            &[
-                BindType::Uniform,
-                BindType::BufReadOnly,
-                BindType::BufReadOnly,
-                BindType::BufReadOnly,
-                BindType::Image(ImageFormat::Rgba8),
-                BindType::ImageRead(ImageFormat::Rgba8),
-                BindType::ImageRead(ImageFormat::Rgba8),
-            ],
-        )?,
-        _ => {
-            engine.add_shader(
+
+    let fine_resources = [
+        BindType::Uniform,
+        BindType::BufReadOnly,
+        BindType::BufReadOnly,
+        BindType::BufReadOnly,
+        BindType::Image(ImageFormat::Rgba8),
+        BindType::ImageRead(ImageFormat::Rgba8),
+        BindType::ImageRead(ImageFormat::Rgba8),
+        // Mask LUT buffer, used only when MSAA is enabled.
+        BindType::BufReadOnly,
+    ];
+    let [fine_area, fine_msaa8, fine_msaa16] = {
+        let aa_support = &options.antialiasing_support;
+        let aa_modes = [
+            (aa_support.area, None),
+            (aa_support.msaa8, Some(("fine_msaa8", "msaa8"))),
+            (aa_support.msaa16, Some(("fine_msaa16", "msaa16"))),
+        ];
+        let mut pipelines = [None, None, None];
+        for (i, (enabled, msaa_info)) in aa_modes.iter().enumerate() {
+            if !enabled {
+                continue;
+            }
+            let (range_end_offset, label, aa_config) = match *msaa_info {
+                Some((label, config)) => (0, label, Some(config)),
+                None => (1, "fine_area", None),
```

Comment:
(I had a big comment written out here, which GitHub mobile helpfully deleted)

But broadly, I was wondering if all of this should be inlined into aa_mode, i.e. make it a `(bool, usize, &str, Option<&str>)`.

If we were doing that, I'd then replace the usize offset with the pipeline resources, so fine_area would be:
`(support.area, &fine_resources[..7], "fine_area", None)`

(Where 7 might not be the right number, but trying to check it lead to the deletion of the previous version of this comment)

Reply:
Done. I just folded the offset into the tuple to keep the column width manageable.